std/btree.asm

```diff
@@ -0,0 +1,153 @@
+use std::utils::Option;
+
+/// This is a BTree, which means that each node has
+/// up to `max_items` items and a number of children
+/// that is one more than the number of items.
+/// The items are key-value-pairs sorted by key and
+/// the key of the i-th item is greater than all keys
+/// in children[i] and less than all keys in children[i + 1].
+/// Leaf nodes do not have children.
+enum BTree<K, V> {
+    Inner((K, V)[], BTree<K, V>[]),
+    Leaf((K, V)[]),
+}
+
+// Rust actually uses 11, but I have the impression that longer
+// arrays do not have that much of an advantage here.
+let max_items = 5;
+
+enum CmpResult {
+    Less,
+    Equal,
+    Greater,
+}
+
+/// Creates an empty BTree.
+let<K, V> new: -> BTree<K, V> = || BTree::Leaf([]);
+
+/// Returns Some(v) if v is the value corresponding to key `k` or None otherwise.
+let<K, V> search: BTree<K, V>, K, (K, K -> CmpResult) -> Option<V> = internal::search::<K, V>;
+
+/// Inserts the key-value-pair (k, v) into the BTree and returns the updated BTree.
+let<K, V> insert: BTree<K, V>, (K, V), (K, K -> CmpResult) -> BTree<K, V> =
+    |b_tree, (k, v), cmp| match internal::insert(b_tree, (k, v), cmp) {
+        internal::InsertResult::Split(item, left, right) => BTree::Inner([item], [left, right]),
+        internal::InsertResult::Updated(t) => t,
+    };
+
+mod internal {
+    use super::BTree;
+    use super::CmpResult;
+
+    use std::utils::Option;
+    use std::utils::unwrap_or_else;
+    use std::array::find_map;
+    use std::array::find_map_enumerated;
+
+    let <K, V> search: BTree<K, V>, K, (K, K -> CmpResult) -> Option<V> = |b_tree, k, cmp| match b_tree {
+        BTree::Inner(items, children) => match search_in_node(items, k, cmp) {
+            NodeSearchResult::InNode(i) => Option::Some(value_of_item(items[i])),
+            NodeSearchResult::InChild(i) => search(children[i], k, cmp),
+        },
+        BTree::Leaf(items) => find_map(items, |(key, value)| match cmp(k, key) {
+            CmpResult::Equal => Option::Some(value),
+            _ => Option::None,
+        })
+    };
+
+    let<K, V> value_of_item: (K, V) -> V = |(_, value)| value;
+
+    enum NodeSearchResult {
+        InNode(int),
+        InChild(int),
+    }
+
+    let<K1, V, K2> search_in_node: (K1, V)[], K2, (K2, K1 -> CmpResult) -> NodeSearchResult = |items, k, cmp| {
```

Comment:
```suggestion
    let<K, V> search_in_node: (K, V)[], K, (K, K -> CmpResult) -> NodeSearchResult = |items, k, cmp| {
```
Right?

Reply:
Our type checker complains that this is not the most generic type signature.

I'm actually considering changing the type checker to only verify that the signature is compatible instead of verifying that it is the most generic one.